### PR TITLE
feature/search

### PR DESCRIPTION
**Pull request by KostSvitlana**

### CODE REVIEW

File: backend/config.py.example [hunk 1]
```diff
@@ -1,24 +0,0 @@
-# Rename this file to config.py and fill SQLALCHEMY_DATABASE_URI with your's data.
```
Comment (serge30) [inline backend/config.py.example]: This is config template file, please don't remove it.

File: backend/views/views.py [hunk 2]
```diff
@@ -2,14 +2,16 @@
...
 
 from app import app, db
-from forms.forms import LoginForm, UserForm
+from forms.forms import LoginForm, SearchForm, UserForm
 from models.users import Role, User
 
 ROLE_ADMIN = 1
 ROLE_MODERATOR = 2
 ROLE_USER = 3
+MIN_SEARCH_STR = 2
```
Comment (serge30) [inline backend/views/views.py]: Different groups of constants please separate by blank line.

File: backend/views/views.py [hunk 3]
```diff
@@ -34,12 +36,43 @@ def admin():
...
 
-@app.route('/userpage')
+@app.route('/userpage', methods=['GET'])
 @admin_permissions
 def user_page():
+    form = SearchForm(request.form)
     users = db.session.query(User, Role).filter(
         User.role_id == Role.id).order_by(User.id).all()
-    return render_template('user_page.html', users=users)
+    if request.args.get('search') != "" and request.args.get('search') >= MIN_SEARCH_STR:
```
Comment (serge30) [inline backend/views/views.py]: Please use `form.validate()` method. In that case you will check search `search` field to min length.

File: backend/views/views.py [hunk 4]
```diff
@@ -34,12 +36,43 @@ def admin():
...
+    # if form.validate_on_submit():
```
Comment (serge30) [inline backend/views/views.py]: Please do not commit commented lines of code.

File: backend/views/views.py [hunk 5]
```diff
@@ -34,12 +36,43 @@ def admin():
...
+        key = int(request.args.get('field_by'))
+        search_string = str(request.args.get('search'))
+        search_paremeters = []
+        search_users = []
+        for one_string in search_string.split(' '):
```
Comment (serge30) [inline backend/views/views.py]: With no parameters [`.split()`](https://docs.python.org/2.7/library/stdtypes.html#str.split) method splits string by whitespaces, which is more suitable for your case.

File: backend/views/views.py [hunk 6]
```diff
@@ -34,12 +36,43 @@ def admin():
...
+            conditions = {1:"users.name LIKE  '%s'" %(search_paremeter),
+                          2:"users.alias LIKE  '%s'" %(search_paremeter),
+                          3:"users.email LIKE  '%s'" %(search_paremeter)}
+            conditions[4] = ' '.join([conditions[1], "OR", conditions[2]])
+            conditions[5] = ' '.join([conditions[2], "OR", conditions[3]])
+            conditions[6] = ' '.join([conditions[1], "OR", conditions[3]])
+            conditions[7] = ' '.join([conditions[1], "OR", conditions[2], "OR", conditions[3]])
+            condition = conditions.get(key)
+            results = db.session.query(User, Role).filter(and_(
+                User.role_id == Role.id, str(condition))).distinct(User.id).order_by(User.id).all()
```
Comment (serge30) [inline backend/views/views.py]: You performs here one query to DB for each part of search string, which adds load to DB. Try to combine all filter expressions in one and perform one query.

File: backend/views/views.py [hunk 7]
```diff
@@ -34,12 +36,43 @@ def admin():
...
+    # if form.validate_on_submit():
+        key = int(request.args.get('field_by'))
+        search_string = str(request.args.get('search'))
+        search_paremeters = []
+        search_users = []
+        for one_string in search_string.split(' '):
+            if len(one_string) >= MIN_SEARCH_STR:
+                search_paremeters.append(''.join(["%", one_string, "%"]))
+        for search_paremeter in search_paremeters:
+            conditions = {1:"users.name LIKE  '%s'" %(search_paremeter),
```
Comment (serge30) [inline backend/views/views.py]: It's not good idea to use strings to form DB query as it may cause SQL-Injection vulnerability. Please use SQLAlchemy's built in [functions](http://docs.sqlalchemy.org/en/latest/orm/tutorial.html#common-filter-operators) for this.

For example this line may be re-written as:
```
{1: User.name.like(search_parameter),
```
`OR` statements below as `or_(conditions[1], conditions[2])`

File: backend/views/views.py [hunk 8]
```diff
@@ -34,12 +36,43 @@ def admin():
...
+            conditions[5] = ' '.join([conditions[2], "OR", conditions[3]])
+            conditions[6] = ' '.join([conditions[1], "OR", conditions[3]])
+            conditions[7] = ' '.join([conditions[1], "OR", conditions[2], "OR", conditions[3]])
+            condition = conditions.get(key)
+            results = db.session.query(User, Role).filter(and_(
+                User.role_id == Role.id, str(condition))).distinct(User.id).order_by(User.id).all()
+            for user in results:
+                if not user in search_users:
+                    search_users.append(user)
+        search_users.sort(key=lambda user: user[0].id)
```
Comment (serge30) [inline backend/views/views.py]: As you will have one DB query, sorting should be performed by SQLAlchemy methods.

File: backend/views/views.py [hunk 9]
```diff
@@ -34,12 +36,43 @@ def admin():
...
+        if search_users == []:
```
Comment (serge30) [inline backend/views/views.py]: `if not search_users:` is enough for this.

File: backend/views/views.py [hunk 10]
```diff
@@ -34,12 +36,43 @@ def admin():
...
+            flash("Search didn`t give result")
+            return redirect(url_for('user_page'))
```
Comment (serge30) [inline backend/views/views.py]: I'm not sure redirect to page with all users is needed, I'd rather show empty list with message that search din't return anything.

File: backend/views/views.py [hunk 11]
```diff
@@ -34,12 +36,43 @@ def admin():
     return render_template('admin_page.html')
 
 
-@app.route('/userpage')
+@app.route('/userpage', methods=['GET'])
 @admin_permissions
 def user_page():
+    form = SearchForm(request.form)
     users = db.session.query(User, Role).filter(
         User.role_id == Role.id).order_by(User.id).all()
```
Comment (serge30) [inline backend/views/views.py]: This query should by only performed when user doesn't provide search query.

**Review by serge30 — CHANGES_REQUESTED** [review]

File: backend/views/views.py (hunk 3 above)
Comment (KostSvitlana) [reply backend/views/views.py]: This method does not work too, maybe problem in form. I don`t know

Comment (serge30) [reply backend/views/views.py]: The root of cause of the issue is that you create WTForm object using `request.form`, but this variable is only available for POST requests, and for GET ones you should use `request.args`. So init statement should be like:
```
form = SearchForm(request.args)
```

Comment (KostSvitlana) [reply backend/views/views.py]: I found an error "CSRF token is missing". When in config.py "WTF_CSRF_ENABLED = False"
method form.validate() is work

Comment (NewIvanIvanov) [reply backend/views/views.py]: add {{ form.hidden_tag() }} into your form

Comment (serge30) [reply backend/views/views.py]: There is no sense to protect search forms from CSRF attacks, you may [disable CSRF](https://flask-wtf.readthedocs.io/en/stable/form.html#module-flask_wtf) protection for `SearchForm` by providing `csrf_enabled=False` parameter:
```
form = SearchForm(request.args, csrf_enabled=False)
```

Comment (KostSvitlana) [reply backend/views/views.py]: Finally it works :) thanks

File: backend/views/views.py [hunk 12]
```diff
@@ -34,12 +37,40 @@ def admin():
...
+                          2:User.alias.like(search_paremeter),
+                          3:User.email.like(search_paremeter)}
+            conditions[4] = or_(conditions[1], conditions[2])
+            conditions[5] = or_(conditions[2], conditions[3])
+            conditions[6] = or_(conditions[1], conditions[3])
+            conditions[7] = or_(conditions[1], conditions[2], conditions[3])
+            condition_list.append(conditions.get(key))
+        condition = or_(condition_list[x] for x in range(len(condition_list)))
+        search_users = db.session.query(User, Role).filter(and_(
+            User.role_id == Role.id, condition)).distinct(User.id).order_by(User.id).all()
```
Comment (serge30) [inline backend/views/views.py]: I believe `distinct()` is not needed anymore.

File: backend/config.py.example [hunk 13]
```diff
@@ -21,4 +21,4 @@ class ProductionConfig(Config):
 
 
 class DevelopmentConfig(Config):
-    DEBUG = True
+    DEBUG = True
```
Comment (serge30) [inline backend/config.py.example]: There is no changes to this file, please revert this changes.

File: backend/views/views.py [hunk 14]
```diff
@@ -34,12 +37,40 @@ def admin():
...
 @admin_permissions
 def user_page():
-    users = db.session.query(User, Role).filter(
-        User.role_id == Role.id).order_by(User.id).all()
-    return render_template('user_page.html', users=users)
+    form = SearchForm(request.args, csrf_enabled=False)
+    if  form.validate():
+        key = int(request.args.get('field_by'))
+        search_string = str(request.args.get('search'))
+        search_paremeters = []
```
Comment (serge30) [inline backend/views/views.py]: typo, should be `search_parameters`

File: backend/views/views.py [hunk 15]
```diff
@@ -34,12 +37,40 @@ def admin():
...
+        condition_list = []
+        for one_string in search_string.split():
+            if len(one_string) >= MIN_SEARCH_STR:
+                search_paremeters.append(''.join(["%", one_string, "%"]))
```
Comment (serge30) [inline backend/views/views.py]: `''.join(["%", one_string, "%"])` is a bit ugly, but `'%%%s%%' % one_string` is confusing too. So I'd recommend to use `'%{}%'.format(one_string)` here.

File: backend/views/views.py [hunk 16]
```diff
@@ -34,12 +37,40 @@ def admin():
...
+        for search_paremeter in search_paremeters:
+            conditions = {1:User.name.like(search_paremeter),
+                          2:User.alias.like(search_paremeter),
+                          3:User.email.like(search_paremeter)}
+            conditions[4] = or_(conditions[1], conditions[2])
+            conditions[5] = or_(conditions[2], conditions[3])
+            conditions[6] = or_(conditions[1], conditions[3])
+            conditions[7] = or_(conditions[1], conditions[2], conditions[3])
+            condition_list.append(conditions.get(key))
+        condition = or_(condition_list[x] for x in range(len(condition_list)))
```
Comment (serge30) [inline backend/views/views.py]: It looks very starnge. Probably you meant:
```
condition = or_(condition_list)
```

File: backend/views/views.py [hunk 17]
```diff
@@ -34,12 +37,40 @@ def admin():
...
+            if len(one_string) >= MIN_SEARCH_STR:
+                search_paremeters.append(''.join(["%", one_string, "%"]))
+        for search_paremeter in search_paremeters:
+            conditions = {1:User.name.like(search_paremeter),
+                          2:User.alias.like(search_paremeter),
+                          3:User.email.like(search_paremeter)}
+            conditions[4] = or_(conditions[1], conditions[2])
+            conditions[5] = or_(conditions[2], conditions[3])
+            conditions[6] = or_(conditions[1], conditions[3])
+            conditions[7] = or_(conditions[1], conditions[2], conditions[3])
```
Comment (serge30) [inline backend/views/views.py]: This part has a low readability, I'd recommend to re-factor it to smth like:
```
user_search = User.name.like(search_parameter)
alias_search = User.alias.like(search_parameter)
email_search = User.email.like(search_parameter)

conditions = [
    user_search,
    alias_search,
    email_search,
    or_(user_search, alias_search),
    ...
]
```

**Review by serge30 — CHANGES_REQUESTED** [review]

File: backend/views/views.py (hunk 16 above)
Comment (KostSvitlana) [reply backend/views/views.py]: I got error "ArgumentError: SQL expression object or string expected, got object of type <type 'list'> instead" therefore I wrote so.

File: backend/views/views.py [hunk 18]
```diff
@@ -34,12 +37,45 @@ def admin():
...
+    form = SearchForm(request.args, csrf_enabled=False)
+    if  form.validate():
+        key = int(request.args.get('field_by'))
+        search_string = str(request.args.get('search'))
+        search_parameters = []
+        condition_list = []
+        for one_string in search_string.split():
+            if len(one_string) >= MIN_SEARCH_STR:
+                search_parameters.append('%{}%'.format(one_string))
+        for search_parameter in search_parameters:
```
Comment (serge30) [inline backend/views/views.py]: Those to loops might be combined in one:
```
for one_string in search_string.split():
    if len(one_string) < MIN_SEARCH_STR:
        continue
    search_parameter = '%{}%'.format(one_string)
    ...
```

File: backend/views/views.py [hunk 19]
```diff
@@ -2,15 +2,18 @@
...
-from forms.forms import LoginForm, UserForm
+from forms.forms import LoginForm, SearchForm, UserForm
 from models.users import Role, User
 
 ROLE_ADMIN = 1
 ROLE_MODERATOR = 2
 ROLE_USER = 3
 
+MIN_SEARCH_STR = 2
+
```
Comment (serge30) [inline backend/views/views.py]: PEP8. 2 blank lines.

File: backend/forms/forms.py [hunk 20]
```diff
@@ -54,3 +54,28 @@ class LoginForm(FlaskForm):
     email = StringField('login', validators=[Email()])
     password = PasswordField('password', validators=[DataRequired()])
     submit_button = SubmitField('Login')
+
```
Comment (serge30) [inline backend/forms/forms.py:101]: PEP8. 2 blank lines.

**Review by serge30 — CHANGES_REQUESTED** [review]

File: backend/views/views.py [hunk 21]
```diff
@@ -34,12 +38,44 @@ def admin():
...
+                name_search,
+                alias_search,
+                email_search,
+                or_(name_search, alias_search),
+                or_(alias_search, email_search),
+                or_(email_search, name_search),
+                or_(name_search, alias_search, email_search)
+                ]
+            condition_list.append(conditions[key])
+        condition = or_(condition_list[x] for x in range(len(condition_list)))
```
Comment (serge30) [inline backend/views/views.py]: So try `condition = or_(*condition_list)`. Anyway I have no idea why your expression works :)

Comment (KostSvitlana) [reply backend/views/views.py]: I experimented with it for a long time:) It is work

**Review by serge30 — APPROVED** [review]
Now it looks good to me. Approved.

**Comment by serge30** [comment]
Please resolve conflicts and merge it to `develop`.
